crates/bevy_sprite/src/tilemap_chunk/mod.rs

```diff
@@ -0,0 +1,124 @@
+use crate::MeshMaterial2d;
+use bevy_app::{App, Plugin, Update};
+use bevy_asset::{Assets, Handle, RenderAssetUsages};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    component::Component,
+    observer::Trigger,
+    query::Changed,
+    system::{Commands, Query, ResMut},
+    world::OnAdd,
+};
+use bevy_image::Image;
+use bevy_math::{primitives::Rectangle, UVec2};
+use bevy_render::{
+    mesh::{Mesh, Mesh2d},
+    render_resource::{Extent3d, TextureDimension, TextureFormat},
+};
+
+mod tilemap_chunk_material;
+
+pub use tilemap_chunk_material::*;
+
+/// Plugin that handles the initialization and updating of tilemap chunks.
+/// Adds systems for processing newly added tilemap chunks and updating their indices.
+pub struct TilemapChunkPlugin;
+
+impl Plugin for TilemapChunkPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_observer(on_add_tilemap_chunk)
+            .add_systems(Update, update_tilemap_chunk_indices);
+    }
+}
+
+/// A component representing a chunk of a tilemap.
+/// Each chunk is a rectangular section of tiles that is rendered as a single mesh.
+#[derive(Component, Clone, Debug, Default)]
+#[require(Mesh2d, MeshMaterial2d<TilemapChunkMaterial>, TilemapChunkIndices)]
+pub struct TilemapChunk {
+    /// The size of the chunk in tiles
+    pub chunk_size: UVec2,
+    /// The size to use for each tile, not to be confused with the size of a tile in the tileset image.
+    /// The size of the tile in the tileset image is determined by the tileset image's dimensions.
+    pub tile_display_size: UVec2,
+    /// Handle to the tileset image containing all tile textures
+    pub tileset: Handle<Image>,
+}
+
+/// Component storing the indices of tiles within a chunk.
+/// Each index corresponds to a specific tile in the tileset.
+#[derive(Component, Clone, Debug, Default, Deref, DerefMut)]
+pub struct TilemapChunkIndices(pub Vec<Option<u32>>);
+
+fn on_add_tilemap_chunk(
+    trigger: Trigger<OnAdd, TilemapChunk>,
+    mut commands: Commands,
+    mut query: Query<(&TilemapChunk, &mut TilemapChunkIndices)>,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<TilemapChunkMaterial>>,
+    mut images: ResMut<Assets<Image>>,
+) {
+    let (
+        TilemapChunk {
+            chunk_size,
+            tile_display_size,
+            tileset,
+        },
+        mut indices,
+    ) = query.get_mut(trigger.target()).unwrap();
+
+    // Ensure the indices vec is the same size as the chunk
+    indices.resize((chunk_size.x * chunk_size.y) as usize, None);
+
+    let indices_image = Image::new(
+        Extent3d {
+            width: chunk_size.x,
+            height: chunk_size.y,
+            depth_or_array_layers: 1,
+        },
+        TextureDimension::D2,
+        indices_image_data(indices.0.clone()),
+        TextureFormat::R32Uint,
+        RenderAssetUsages::default(),
+    );
+
+    commands.entity(trigger.target()).insert((
+        Mesh2d(meshes.add(Rectangle::from_size(
```

Comment:
Does this really need a new mesh handle for each chunk? I assume most users will have the same chunk_size + tile_display_size. You could store the handle in a map with the size as the key. This would only be an issue if a user creates thousands of chunks with unique sizes, but for the average case of using the same size everywhere it will perform better.